Refactor clustering (#1728)
emqtt/emq-docs-cn: en_US/deploy/cluster/create-cluster.md

# Create an EMQX Cluster

> EMQX supports creating clusters manually and automatically. There are 2 key concepts before creating a cluster, node names and node discovery. 

Author: ```suggestion
With EMQX cluster, you can enjoy the benefits of fault tolerance and high availability by allowing the cluster to continue operating even if one or more nodes fail. You can create an EMQX cluster either manually or automatically. Before the actual clustering step, let's first get familiar with 2 key concepts, node names and node discovery. 
```

## Concepts

### Node Names

> Before starting the cluster creation step, let's first get familiar with the concept of node names in EMQX. EMQX nodes are identified by their names. A node name consists of two parts, node name and host, separated with `@`, for example, `emqx@s1.emqx.io`. The host part must either be the IP address or FQDN (`myhost.example.domain`), for example:

Reviewer: Nit, I think it's good to explain acronyms before first use:
```suggestion
Before starting the cluster creation step, let's first get familiar with the concept of node names in EMQX. EMQX nodes are identified by their names. A node name consists of two parts, node name and host, separated with `@`, for example, `emqx@s1.emqx.io`. The host part must either be the IP address or fully qualified domain name (FQDN), such as `myhost.example.tld`, for example:
```

- For EMQX node deployed on server `s1.emqx.io`, the node name should be `emqx@s1.emqx.io`; 
- If this server has a static IP (`192.168.0.10`), the node name should be `emqx@192.168.0.10`. 

::: tip Tip

> EMQX node names are immutable, as they are baked into the database schema and data files. Therefore, it is recommended to use static FQDNs for EMQX node names.

Reviewer: I think the part about preferring FQDNs to static IPs is very important, since we have had incidents related to this.

:::

### Node Discovery

EMQX's clustering feature is based on the [Ekka](https://github.com/emqx/ekka) library, a cluster management library developed for Erlang/OTP applications.

One of the crucial steps for EMQX clustering is node discovery, which enables individual EMQ X nodes with different IP addresses or locations to discover and communicate with each other. EMQX supports multiple node discovery strategies:

| Strategy | Description                               |
| -------- | ----------------------------------------- |
| `manual` | Create a cluster through manual command   |
| `static` | Create a cluster using a static node list |
| `mcast`* | Create a cluster through UDP multicast    |

> | `dns`    | Create a cluster using DNS A records      |

Reviewer 2: DNS A and DNS SRV

> | `dns`    | Create a cluster using DNS A records      |

Author: ```suggestion
| `dns`    | Create a cluster using DNS A and SRV records      |
```

| `etcd`   | Create a cluster via etcd                 |
| `k8s`    | Create a cluster via Kubernetes service   |

[^*]: The multicast discovery strategy has been deprecated and will be removed in future releases.

## Before You Begin

Before creating an EMQX cluster, the following prerequisites should first be met:

1. All nodes are set with a unique node name in the format of `name@host`, where host must be an IP address or fully qualified domain name (FQDN). For more information on configuring node names, see [Configure Node Names](#configure-node-names). 

2. If there is a firewall or security group between nodes, ensure the cluster communication port has been opened. For details, see [Intra-cluster communication port](./security.md).

> 3. For security concerns, you should change the default cookie settings to `emqxsecretcookie` in `emqx.conf` on all nodes to join the cluster. Note: All nodes to join the cluster should use the same security cookie. For details about the magic cookie used, see [Distributed Erlang - Security](https://www.erlang.org/doc/reference_manual/distributed.html#security). 

Reviewer 2: Change to a secret cookie

   ```
   node {

>      cookie = "emqxsecretcookie"

Reviewer 2: Remind the user explicitly again
```suggestion
     cookie = "<a secret cookie>"
```

   }
   ```

:::

Now you can begin your journey to create an EMQX cluster. 

## Configure Node Names

Before creating a cluster, you need first to name the nodes to join the cluster. Suppose you want to create a cluster for 2 nodes deployed in `s1.emqx.io` and `s2.emqx.io` respectively, you can follow the steps below to create the cluster. 

Configure the node name in the `emqx.conf` configuration file of the 1st node, for example:

```bash
node.name = emqx@s1.emqx.io
# or

> node.name = emqx@192.168.0.10

Reviewer: Perhaps leave out static IPs from the examples, since we generally discourage their use in the node names.

Author: agree, so i will remove this line

> node.name = emqx@192.168.0.10

Author: ```suggestion

```

```

You can also override the node name with an environment variable:

```bash
env EMQX_NODE__NAME='emqx@s1.emqx.io' ./bin/emqx start
```

Repeat the above step for the other node to join the cluster. 

Now you have named 2 nodes to join the cluster, `emqx@s1.emqx.io` and `emqx@s2.emqx.io`. You can create a cluster either manually or automatically. 

## Manual Clustering

### Set Node Discovery Strategy

EMQX supports creating clusters manually and automatically. This section will introduce the manual clustering feature in EMQX. 

Set node discovery strategy

Manual clustering is the method to configure an EMQX cluster by manually specifying which nodes should be part of the cluster. By default, EMQX adopts a manual clustering strategy, which can also be set in `emqx.conf`. Besides, you also need to configure the default 

```bash
cluster {
    ## options: manual | static | dns | etcd | K8s
    discovery_strategy  =  manual
}
```

> This approach provides users with greater control and flexibility over the cluster configuration and more security as you can restrict which nodes can join the cluster. 

Reviewer: > and more security as you can restrict which nodes can join the cluster.

Not really... Manual discovery strategy is essentially just the absence of automatic service discovery. It does not impose any additional restrictions or security checks. Any node with the right cookie still can join the a cluster, it just requires `emqx_ctl cluster join ...` command by the operator.

Author: so I think I will just remove this benefit part.

> This approach provides users with greater control and flexibility over the cluster configuration and more security as you can restrict which nodes can join the cluster. 

Author: ```suggestion

```

> :::tip

Author: ```suggestion

```

> Manual clustering is not supported in the Core-Replica architecture.

Author: ```suggestion

```

> :::

Reviewer: This issue has since been fixed, so this tip can be removed.

Author: removed, thanks

> :::

Author: ```suggestion

```

### Configure Nodes to Join a Cluster

> After the nodes are started,  you can run the `cluster join` command on the node that you want to join the cluster. For example, you want `emqx@s2.emqx.io` to join  `emqx@s1.emqx.io`, run the command below on `emqx@s2.emqx.io`：

Reviewer: Perhaps it makes sense to include `emqx_ctl` here to avoid confusion?

```suggestion
After the nodes are started,  you can run the `emqx_ctl cluster join` command on the node that you want to join the cluster. For example, you want `emqx@s2.emqx.io` to join  `emqx@s1.emqx.io`, run the command below on `emqx@s2.emqx.io`：
```

> <!--should we add how to start the nodes or how to confirm the nodes are started?-->

Reviewer 2: ```
emqx ping
```

```bash
$ ./bin/emqx_ctl cluster join emqx@s1.emqx.io

Join the cluster successfully.
Cluster status: [{running_nodes,['emqx@s1.emqx.io','emqx@s2.emqx.io']}]
```

:::tip

1. This command must run on the node to join the cluster, that is, as a **request** rather than **invite**.

2. After `emqx@s2.emqx.io` joins `emqx@s1.emqx.io` to form a cluster, it will clear the local data and synchronize the data in `emqx@s1.emqx.io`.

> 3. If `emqx@s2.emqx.io`  wants to join another cluster, it must first leave the current cluster. 

Reviewer 2: how to leave
```
./bin/emqx_ctl cluster leave
```

   :::

Now you have successfully created a cluster with two nodes, you can read the [Query Cluster Status](#query-cluster-status), [Manage Cluster Nodes](#manage-cluster-nodes), and [Configure Network Protocols](#configure-network-protocols) sections on how to monitor the cluster status and how to manage the cluster. 

## Auto Clustering

> Auto clustering in EMQX is another feature that allows multiple EMQX nodes to form a cluster automatically without manual configuration. Auto clustering simplifies the process of setting up an EMQX cluster and makes it easier to add or remove nodes from the cluster dynamically. It also provides fault tolerance and high availability by allowing the cluster to continue operating even if one or more nodes fail.

Reviewer: > It also provides fault tolerance and high availability by allowing the cluster to continue operating even if one or more nodes fail.

The same applies to the manual cluster as well: manual cluster will also keep working when one or several members fail, as long as at least one member is still running... 
And manual cluster recovers automatically as well, as soon as the faulty nodes come back to live (given that their persistent data was not wiped); there is no need to manually re-join them to the cluster after the initial command. 

Perhaps this sentence be moved to the top of the chapter, since fault-tolerance is not strictly a property of automatic clustering, but clustering in general.

Author: thanks, moved this sentence to the very beginning of this chapter.

> Auto clustering in EMQX is another feature that allows multiple EMQX nodes to form a cluster automatically without manual configuration. Auto clustering simplifies the process of setting up an EMQX cluster and makes it easier to add or remove nodes from the cluster dynamically. It also provides fault tolerance and high availability by allowing the cluster to continue operating even if one or more nodes fail.

Author: ```suggestion
Auto clustering in EMQX is another feature that allows multiple EMQX nodes to form a cluster automatically without manual configuration. Auto clustering simplifies the process of setting up an EMQX cluster and makes it easier to add or remove nodes from the cluster dynamically.
```

#### Configure DNS Services

Most public cloud services have DNS services. After assigning a domain name, you only need to add the IP address of each EMQX node to the A record of this domain to finish the configuration.

:::tip

If EMQX is deployed in a private cloud or internal network, you will need to deploy your own DNS system, for example, with software [BIND](https://www.isc.org/bind/).

:::

### Configure Autocluster by DNS Records

After the DNS service is ready, you can add all nodes to join the cluster in `emqx.conf` with the `cluster.dns` configuration item:

**Example code:**

```bash
cluster {
    discovery_strategy = dns
    dns {
        name = "localhost"
        ## support DNS A record and DNS SRV record 
        record_type = a
    }
}
```

Where, 

- `discovery_strategy` is the node discovery strategy, set it to `dns`
- `cluster.dns.name` is a a string, input the localhost
-  `cluster.dns.record_type` is a enum, optional value: `a` or `srv`

> After all nodes are started, the cluster will be automatically established.

Reviewer 2: Question: Is there a requirement for the start-up order?

Author: @ieQu1 could you please check the question @wivwiv left above?

- `cluster.K8s.namespace` is the Kubernetes namespace, it is a string object, default: `default`

Start all nodes one by one after the configuration, and the cluster will be automatically established.

::: tip

When working EMQX autocluster on Kubernetes, [Calico](https://kubernetes.io/docs/tasks/administer-cluster/network-policy-provider/calico-network-policy/) rather than Fannel plugin is recommended. 
:::

After the cluster is succefully created, you can refer to the section below on how to monitor the cluster status and how to manage cluster nodes. 

## Query Cluster Status

Run the command below on any cluster node to query the cluster status:

```bash
$ ./bin/emqx_ctl cluster status

Cluster status: [{running_nodes,['emqx@s1.emqx.io','emqx@s2.emqx.io']}]
```

## Manage Cluster Nodes

You can remove a node from a cluster with `cluster leave` or ``cluster force-leave``:

When an EMQX node issues the `cluster leave` command, it notifies the other nodes in the cluster that it intends to leave, and it stops participating in cluster operations, it will complete any ongoing tasks before leaving.

> When an EMQX node issues the `cluster force-leave` command, a node will be forcefully removed from a cluster. This command is typically used when a node fails or becomes unresponsive. 

Reviewer: Force leave command is a bit confusing, since, opposite to `leave` command, it removes a remote node from the cluster. So perhaps it can be emphasized like this:  
 
```suggestion
When an EMQX node issues the `cluster force-leave <node@host>` command, the target node will be forcefully removed from a cluster. This command is typically used when a node fails or becomes unresponsive. 
```
